File: 0120-triangle/0120-triangle.cpp
class Solution {
public:
    int solve(int i, int j, int m, vector<vector<int>>& dp, vector<vector<int>>& t)
    {
        
        if(i==m-1)
           return t[i][j];
        if(dp[i][j]!=-1)
           return dp[i][j];
        return dp[i][j]=t[i][j]+ min(solve(i+1,j,m,dp,t), solve(i+1,j+1,m,dp,t));


    }
    int minimumTotal(vector<vector<int>>& t) {
        int m = t.size();
        vector<vector<int>>dp(m+1,vector<int>(m+1,-1));
        return solve(0,0,m,dp,t);
    }
};

File: 0120-triangle/0120-triangle.java
class Solution {
    int mini(int i,int j,int dp[][],List<List<Integer>> a,int n){
 
        if(i==n-1)
        return a.get(i).get(j);

        if(dp[i][j]!=-1)
        return dp[i][j];
        
        int up=a.get(i).get(j)+mini(i+1,j,dp,a,n);
        int down=a.get(i).get(j)+mini(i+1,j+1,dp,a,n);
        return dp[i][j]=Math.min(up,down);

    }

    public int minimumTotal(List<List<Integer>> triangle) {
        int n=triangle.size();
        System.out.println(n);
        int dp[][]=new int[n][n];
        for(int row[]:dp)
        Arrays.fill(row,-1);

        return mini(0,0,dp,triangle,n);
        
    }
}
